### PR TITLE
Refactor operations to eliminate duplicate code between EVM versions

### LINKED ISSUE #6902: `HederaEvmCreateOperation`

### LINKED ISSUE #6903: `HederaEvmCreate2Operation`

### LINKED ISSUE #6928: Refactor `HederaBalanceOperation` to use `HederaAddresses`
The current family of `HederaOperationUtil`, `HederaOperationUtilV038`, `HederaEvmOperationsUtil`, and `HederaEvmOperationsUtilV038` helpers do an good of eliminating duplication.

However, their method signatures rely heavily on `java.util.function` types; e.g. [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/contracts/operations/HederaEvmOperationsUtilV038.java#L48) we see a sequence of 6 **(!)** standard functional interfaces, which hurts readability.

We would prefer to have tiny bit of structural duplication in the `Operation`s that call these utilities to let the reader's eye move more naturally through the logic and how it incorporates Hedera-specific address checks.

### LINKED ISSUE #6950: `HederaBalanceOperationV038`

### LINKED ISSUE #6954: `HederaEvmChainIdOperation`

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Refactors most customized `Operation`s to vary based on injected `AddressChecks` and `FeatureFlags` (which together will encapsulate the varying behavior of the EVM version for release).
 - Closes #6950 
 - Closes #6954
 - Closes #6928
 - Closes #6903 
 - Closes #6902 
 - Still in-progress: `CustomCallCodeOperation`, `CustomSelfDestructOperation`, `CustomSLoadOperation`, `CustomLogOperation`.

**Notes to reviewer**:
The main constraints relative to `mono-service` are:
  1. The `ContractService` can no longer directly mutate any account, token, or token association property (`nonce`, `deleted`, `balance`, `supply`, etc.) and must use a "dispatch" facility provided by something called the `HandleContext`.
  2. The `ContractService` can no longer directly create or manage records.

The main goals relative to `mono-service` are:
 1. Mirror the Besu `TransactionProcessor` pattern much more closely, where given a `ProtocolSpec`, a transaction can be handled in a single call with a bit of context (e.g. current `gasPrice`, `BlockValues`) and a choice of EVM version.
 2. Encapsulate the EVM version-specific behavior in a few well-defined interfaces (`FeatureFlags`, `AccountChecks`) **instead of** having different `Operation` implementations for each versions.
 3. Have a cleaner boundary between Besu types and Hedera types.
 4. Make the final module much more reusable for mirror nodes.

### CODE REVIEW

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/bni/VerificationStrategy.java [hunk 1]
```diff
@@ -74,6 +74,8 @@ enum KeyRole {
      * @return an amended CryptoTransfer, or null if no amendment is necessary
      */
     @Nullable
-    CryptoTransferTransactionBody maybeAmendTransfer(
-            @NonNull CryptoTransferTransactionBody transfer, List<Long> invalidSignerNumbers);
+    default CryptoTransferTransactionBody maybeAmendTransfer(
+            @NonNull CryptoTransferTransactionBody transfer, List<Long> invalidSignerNumbers) {
+        throw new UnsupportedOperationException();
```
Comment (hendrikebbers) [inline hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/bni/VerificationStrategy.java:79]: Please add a message

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.contract.impl.annotations;
+
+import javax.inject.Qualifier;
+
+@Qualifier
```
Comment (hendrikebbers) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java:21]: `@Target`, `@Retention` and `@Documented` missing

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+public @interface ServicesV030 {}
```
Comment (hendrikebbers) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java:22]: Can you add some documenation

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java (hunk 2 above)
Comment (hendrikebbers) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java:21]: Same for all other new annotations in the package

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java (hunk 3 above)
Comment (hendrikebbers) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java:22]: Same for all other new annotations in the package

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ServiceModule.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    Map<Address, PrecompiledContract> bindHederaPrecompiles();
+
+    @Provides
+    @Singleton
+    static Map<String, TransactionProcessor> provideTransactionProcessors(
+            @ServicesV030 @NonNull final TransactionProcessor v030Processor,
+            @ServicesV034 @NonNull final TransactionProcessor v034Processor,
+            @ServicesV038 @NonNull final TransactionProcessor v038Processor) {
+        return Map.ofEntries(
+                entry("v0.30", v030Processor), entry("v0.34", v034Processor), entry("v0.38", v038Processor));
```
Comment (hendrikebbers) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ServiceModule.java:61]: Can we add constants for the versions?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomCallOperation.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    private final FeatureFlags featureFlags;
+    private final AddressChecks addressChecks;
+
+    public CustomCallOperation(
+            @NonNull final FeatureFlags featureFlags,
+            @NonNull final GasCalculator gasCalculator,
+            @NonNull final AddressChecks addressChecks) {
+        super(gasCalculator);
+        this.featureFlags = featureFlags;
```
Comment (hendrikebbers) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomCallOperation.java:59]: null checks are missing

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomChainIdOperation.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ * A {@code CHAINID} operation that uses the {@link com.hedera.node.config.data.ContractsConfig} from the
+ * frame context variables to provide the chain id. (The
+ * {@link com.hedera.node.app.service.contract.impl.exec.TransactionProcessor} must always set
+ * the config in the frame context, since in principle it could vary with every transaction.)
+ */
+public class CustomChainIdOperation extends AbstractOperation {
+    private final long cost;
+
+    public CustomChainIdOperation(@NonNull final GasCalculator gasCalculator) {
+        super(0x46, "CHAINID", 0, 1, gasCalculator);
```
Comment (hendrikebbers) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomChainIdOperation.java:41]: `0x46` is a magic number for me. Can you create a constant and document it?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomCreate2Operation.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+
+public class CustomCreate2Operation extends AbstractCustomCreateOperation {
+    private static final Bytes EIP_1014_PREFIX = Bytes.fromHexString("0xFF");
+    private final FeatureFlags featureFlags;
+
+    public CustomCreate2Operation(
+            @NonNull final GasCalculator gasCalculator, @NonNull final FeatureFlags featureFlags) {
+        super(0xF5, "ħCREATE2", 4, 1, gasCalculator);
```
Comment (hendrikebbers) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomCreate2Operation.java:39]: Magic number

File: hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/bni/VerificationStrategy.java (hunk 1 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/bni/VerificationStrategy.java]: Done.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java:35]: Added.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/annotations/ServicesV030.java:36]: Done.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ServiceModule.java (hunk 4 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/ServiceModule.java]: Done.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomCallOperation.java (hunk 5 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomCallOperation.java]: Added.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomChainIdOperation.java (hunk 6 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomChainIdOperation.java]: Moved these opcodes to an `enum`.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomCreate2Operation.java (hunk 7 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomCreate2Operation.java]: Moved to `enum`.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCalculator.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+package com.hedera.node.app.service.contract.impl.exec.gas;
+
+import org.hyperledger.besu.evm.gascalculator.LondonGasCalculator;
+
+/**
+ * Need to clean up and refactor {@link com.hedera.node.app.service.mono.contracts.gascalculator.GasCalculatorHederaV22}
+ * to get its price and exchange rate from the transaction's {@link com.hedera.node.app.spi.meta.bni.Scope}.
+ */
+public class CustomGasCalculator extends LondonGasCalculator {}
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/CustomGasCalculator.java:25]: Should we be extending London or Shandong.? I believe that the ShandongGasCalculator had the update for EIP-3860 limit the size of the initcode

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/AbstractCustomCreateOperation.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+     * used to give specializations the chance to do any Hedera-specific logic.
+     *
+     * @param frame the frame running the successful operation
+     * @param createdAddress the address of the newly created contract
+     */
+    protected abstract void onSuccess(@NonNull MessageFrame frame, @NonNull Address createdAddress);
+
+    @Override
+    public OperationResult execute(@NonNull final MessageFrame frame, @NonNull final EVM evm) {
+        if (!isEnabled(frame)) {
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/AbstractCustomCreateOperation.java:105]: Did this get copied and modified from `AbstractEvmRecordingCreateOperation` or from `AbstractCreateOperation` in besu?  The implementations seems very similar but slightly different.  I'm not sure if the differences are meaningful.

File: hedera-node/hedera-smart-contract-service-impl/build.gradle.kts [hunk 10]
```diff
@@ -28,6 +28,10 @@ dependencies {
     }
 }
 
+val generatedSources = file("build/generated/sources/annotationProcessor/java/main")
```
Comment (hendrikebbers) [inline hedera-node/hedera-smart-contract-service-impl/build.gradle.kts:31]: what is the change in this file used for?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/FeatureFlags.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Whether "implicit creation" of accounts via sending value or targeting a {@code CREATE2} to an EIP-1014 address
+     * should be enabled for the given {@code frame}.
+     *
+     * @param frame the {@link MessageFrame} to check
+     * @return whether implicit creation should be enabled
+     */
+    boolean isImplicitCreationEnabled(@NonNull MessageFrame frame);
+}
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/FeatureFlags.java:42]: Do we need an interface for this or can we get from config ?

**Review by lukelee-sl — APPROVED** [review]
LGTM

**Pull request merged**